### PR TITLE
Instead of storing RRs, store hashes and require proofs from callers

### PR DESCRIPTION
**Pull request by Arachnid**

### CODE REVIEW

**Comment by makoto** [comment]
One test failing
```
   1) should reject entries with expirations in the past

    Events emitted during test:
    ---------------------------

    RRSetUpdated(name: 0x00, rrset: 0x000030000100000e100006010103fd1111000030000100000e100006000004fd1111000030000100000e100006000003fd1112)
```

**Comment by makoto** [comment]
Just ran the test and compared the gas price. Looks like gas price is almost half which is 👍 
```
│         Contract  ·  Method          ·    Min     ·  Max      ·  Avg      ·  # calls     ·  usd (avg)  │
│  BEFORE DNSSEC    ·  submitRRSet     ·    82975  ·   861034  ·   272520  ·          13  ·       0.20  │
│  AFTER  DNSSEC    ·  submitRRSet     ·    83298  ·   476217  ·   152560  ·          16  ·       0.11  │
```

File: contracts/DNSSEC.sol [hunk 1]
```diff
@@ -124,113 +126,114 @@ contract DNSSEC is Owned {
...
-     * @param dnsclass The DNS class (1 = CLASS_INET) of the records being inserted.
      * @param name The name of the RRSET, in DNS label-sequence format.
      * @param input The signed RR set. This is in the format described in section
      *        5.3.2 of RFC4035: The RRDATA section from the RRSIG without the signature
      *        data, followed by a series of canonicalised RR records that the signature
      *        applies to.
      * @param sig The signature data from the RRSIG record.
      */
-    function submitRRSet(uint16 dnsclass, bytes memory name, bytes memory input, bytes memory sig) public {
+    function submitRRSet(bytes memory name, bytes memory input, bytes memory sig, bytes memory proof) public {
```
Comment (makoto) [inline contracts/DNSSEC.sol]: In the test, `input` is mentioned as `data`. Better to use consistent name

File: contracts/DNSSEC.sol [hunk 2]
```diff
@@ -328,17 +338,13 @@ contract DNSSEC is Owned {
...
      * @param keyrdata The RDATA section of the key.
      * @param keytag The keytag of the key.
      * @param algorithm The algorithm ID of the key.
      * @return True if a DS record verifies this key.
      */
-    function verifyKeyWithDS(uint16 dnsclass, bytes memory keyname, bytes memory keyrdata, uint16 keytag, uint8 algorithm) internal view returns (bool) {
-        bytes memory data = rrsets[keccak256(keyname)][DNSTYPE_DS][dnsclass].rrs;
-        if(data.length == 0) return false;
-
+    function verifyKeyWithDS(bytes memory keyname, bytes memory keyrdata, uint16 keytag, uint8 algorithm, bytes memory data) internal view returns (bool) {
```
Comment (makoto) [inline contracts/DNSSEC.sol]: Better to rename `data` to `proof` for consistency

**Comment by makoto** [comment]
If I understand this correctly, you are only saving the hash of rrs and you need to provide the unhashed rrs as a proof of anchors. This means that when you deleteRRS, do we also need to pass the rrs of NSEC as a proof and compare the hash against the saved hash? Are there any possibility that they don't much as some data which we retrieve from DNS has changed (eg: ttl)?

File: truffle.js [hunk 3]
```diff
@@ -19,5 +19,11 @@ module.exports = {
       currency: 'USD',
       gasPrice: 1
     }
+  },
+  solc: {
```
Comment (makoto) [inline truffle.js]: What's this for?

Comment (Arachnid) [reply truffle.js]: It turns on the optimiser.

File: test/enssec.js [hunk 4]
```diff
@@ -331,61 +356,66 @@ contract('DNSSEC', function(accounts) {
...
     assert.equal((await checkPresence(instance, dns.TYPE_TXT, 'b.')), false);
   })
 
   // Test against real record
   it('should accept real DNSSEC records', async function() {
     var instance = await dnssec.deployed();
+    var proof = await instance.anchors();
     for(var rrset of test_rrsets) {
-      var tx = await verifySubmission(instance, rrset[0], "0x" + rrset[1], "0x" + rrset[2]);
+      console.log(rrset[0]);
```
Comment (makoto) [inline test/enssec.js]: Do you want to log this or just a left over debug console?

**Pull request merged**